### PR TITLE
Ablang2 loss

### PR DESCRIPTION
**Pull request by kleinhenz**
Adds ablang2 loss and tests. Supports specifying `designable_positions` which limits the set of residues that the ppl is calculated over.

### CODE REVIEW

File: src/mosaic/losses/ablang2.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+def _load_ablang2():
+    from ablang2.load_model import load_model
+    from jablang import from_torch
+
+    model_pt, tokenizer, _hparams = load_model("ablang2-paired")
+    model_pt.eval()
+    return from_torch(model_pt), tokenizer
+
+
+@lru_cache(maxsize=1)
```
Comment (nboyd) [inline src/mosaic/losses/ablang2.py]: I think it would be nicer to have a non-private `load_ablang2` function without the LRU cache -- it's a small model and I'd trust people to not load multiple times.

File: src/mosaic/losses/ablang2.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+class Ablang2PseudoLikelihood(LossTerm):
+    """Pseudo-likelihood loss using the AbLang2 paired model.
+    Formats the concatenated binder sequence as ``<H>|<L>`` (or ``<H>|`` /
+    ``|<L>`` for single-chain) to match ablang2's input convention, and masks
+    special-token logits before log-softmax.
+    """
+
+    model: Any
+    tokenizer: Any
+    chain_slices: tuple[tuple[str, int, int], ...]  # (part_id, start, stop); part_id must be "H" or "L"
```
Comment (nboyd) [inline src/mosaic/losses/ablang2.py]: I found this a bit confusing. Instead, what if you followed the convention that your sequence is always heavy then light? Then you'd only need a single int here for length(heavy).

File: src/mosaic/losses/ablang2.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    designable_positions: jax.Array | None
```
Comment (nboyd) [inline src/mosaic/losses/ablang2.py:40]: it looks like you're computing the PLL of only the designable positions, which make sense to me. OTOH I think the gradient is going to be identical (if stop_gradient=True, and up to scaling, maybe?) if you precompose this loss with [SetPositions](https://github.com/escalante-bio/mosaic/blob/694f8de4837295ad2515d93f1c1742c9c7cece9f/src/mosaic/losses/transformations.py#L87), which seems like a more natural way to handle this. The value will be different, but I think it's worth it for the simplification. 

It's possible I'm missing a very good reason to do it this way, let me know!

File: src/mosaic/losses/ablang2.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        def special(token):
+            return jax.nn.one_hot(jnp.array([at[token]]), self.vocab_size)
+
+        chain_bounds = {cid: (start, stop) for cid, start, stop in self.chain_slices}
+        parts: list[jax.Array] = []
+        sequence_token_indices = jnp.full(n, -1, dtype=jnp.int32)
+        offset = 0
+
+        for part_id in ("H", "|", "L"):
```
Comment (nboyd) [inline src/mosaic/losses/ablang2.py]: if we have the int property for the chain break I think this becomes much simpler

**Review by nboyd — CHANGES_REQUESTED** [review]
This looks very good, thanks @kleinhenz !

I added a few minor comments/questions about style and intent -- let me know what you think.

By the way, I'm sure you've figured this out but if you're doing multi-chain design you'll need to manually construct features using `target_only_features` with separate chains for H and L. I bet you're already doing this because you're fixing positions (so passing sequences like "QVQLVESGGGLVQPGGSLRLSCAASXXXXXXXXXXXLGWFRQ...," but just wanted to note.

File: src/mosaic/losses/ablang2.py (hunk 3 above)
Comment (kleinhenz) [reply src/mosaic/losses/ablang2.py:40]: I agree that the gradients are the same up to a scaling factor of M/N where M is the number of designable positions and N is the total length of the binder chains. What this does is:

1) avoid the undesirable scaling factor. I don't want the naturalness to matter less for fewer designed residues. (this could go into the loss construction as an explicit rescaling though)
2) make the ppl we save in the aux more informative since the designable positions aren't averaged together with fixed framework residues
3) save some compute since we avoid evaluating on fixed positions that don't contribute to the gradient

Not sure that this is worth the complexity though. It would be nice to have a more composable way of threading this information through the losses but not sure exactly what the shape of that would be.

File: src/mosaic/losses/ablang2.py (hunk 2 above)
Comment (kleinhenz) [reply src/mosaic/losses/ablang2.py]: yes this is much better! updated

File: src/mosaic/losses/ablang2.py (hunk 3 above)
Comment (kleinhenz) [reply src/mosaic/losses/ablang2.py:40]: One other thing to note is this does compose fine with `SetPositions`. See [this test](https://github.com/kleinhenz/mosaic/blob/ablang2_loss/tests/test_ablang2_loss.py#L274).

File: src/mosaic/losses/ablang2.py (hunk 1 above)
Comment (kleinhenz) [reply src/mosaic/losses/ablang2.py]: done

File: src/mosaic/losses/ablang2.py (hunk 3 above)
Comment (nboyd) [reply src/mosaic/losses/ablang2.py:40]: good points! let's merge as is. 

the larger question of how to handle this kind of stuff in general is really interesting. I've thought at bit about allowing losses to return length n arrays and relying on XLA DCE to do stuff like this... but seems hard.

**Comment by nboyd** [comment]
Thanks again @kleinhenz ! Looks great. If you find this works well in an antibody design workflow would be great to add an example script later
